system/jlib/jsocket.cpp

```diff
@@ -445,22 +449,121 @@ extern jlib_decl bool queryKeepAlive(int &time, int &intvl, int &probes)
                     keepAliveTime = keepalive->getPropInt("@time", keepAliveTime);
                     keepAliveInterval = keepalive->getPropInt("@interval", keepAliveInterval);
                     keepAliveProbes = keepalive->getPropInt("@probes", keepAliveProbes);
-                    state = ENABLED;
+                    hasKeepAlive = true;
                 }
+                disableDNSTimeout = expert->getPropBool("@disableDNSTimeout", false);
+                maxDNSThreads = expert->getPropInt("@maxDNSThreads", maxDNSThreads);
+                // could also consider maxDNSThreads==0 as a way to disable DNS timeout ...
+                if (maxDNSThreads < 1)
+                    maxDNSThreads = 1;
             }
-            doKeepAlive = state;
+            expertTCPSettings = INITED;
         }
     }
+}
 
-    if (state == ENABLED)
+extern jlib_decl bool queryKeepAlive(int &time, int &intvl, int &probes)
+{
+    queryTCPSettings();
+    if (hasKeepAlive)
     {
         time = keepAliveTime;
         intvl = keepAliveInterval;
         probes = keepAliveProbes;
+    }
+    return hasKeepAlive;
+}
+
+static bool getAddressInfo(const char *name, unsigned *netaddr, bool okToLogErr);
+
+static CriticalSection queryDNSCS;
+
+class CAddrInfoThreadArgs : public CInterface, implements IInterface
+{
+public:
+    IMPLEMENT_IINTERFACE;
+    StringAttr name;
+    unsigned netaddr[4] = { 0, 0, 0, 0 };
+    bool result = false;
+
+    CAddrInfoThreadArgs(const char *_name) : name(_name) { }
+};
+
+class CAddrPoolThread : public CInterface, implements IPooledThread
+{
+public:
+    IMPLEMENT_IINTERFACE;
+
+    Linked<CAddrInfoThreadArgs> localAddrThreadInfo;
+
+    virtual void init(void *param) override
+    {
+        localAddrThreadInfo.set((CAddrInfoThreadArgs *)param);
+    }
+
+    virtual void threadmain() override
+    {
+        localAddrThreadInfo->result = getAddressInfo(localAddrThreadInfo->name.get(), localAddrThreadInfo->netaddr, false);
+        localAddrThreadInfo->Release();
```

Comment:
You shouldn't need to explicitly call release here - you could call setown rather than set in the init() function.  Probably better is to not call getLink() when start is called, and then you can remove the calls to             addrThreadInfo->Release() in the exception handling.

Reply:
ok, thanks.
I've changed set() to setown() and removed the Release().